### ISSUE TITLE
How to verify when the signature is in a different file than the content?

### ISSUE THREAD

**Issue by linkurzweg** [PeculiarVentures/xmldsigjs#61]
I'm trying to verify some XML files in the browser. I have a file document.xml and a document.sign.xml which contains a reference tag that references the document like this
```
<Reference URI="document.xml">
    <DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
    <DigestValue>...</DigestValue>
</Reference>
```

I am trying to verify the document like this (xml is document.xml and sigXml is document.sign.xml):

```
const parsedDoc = Parse(xml)
const parsedSig = Parse(sigXml)

const signature = parsedSig.getElementsByTagNameNS('http://www.w3.org/2000/09/xmldsig#', 'Signature')

const signedXml = new SignedXml(parsedDoc)
signedXml.LoadXml(signature[0])

try {
  const res = await signedXml.Verify()
  console.log('signature status', res)
} catch (error) {
  console.log(error)
}
```

But I get an error every time that the digest values do not match
` "XMLJS0013: Cryptographic error: Invalid digest for uri 'document.xml'. Calculated digest is ... but the xml to validate supplies digest ...`

If I try to verify the signature with `xmlsec1 --verify document.sign.xml` it says the file is ok:
```
OK
SignedInfo References (ok/all): 1/1
Manifests References (ok/all): 0/0
```

How can I do this properly in the browser?

**Comment by microshine** [comment]
Could you share your files for testing?

**Comment by linkurzweg** [comment]
Sorry, I am not allowed to, because the files contain confidential information :/

**Comment by microshine** [comment]
Perhaps the canonicalization method returns the incorrect result.

Could you run `xmlsec1` application with `--store-references` option? It must print the result of `<dsig:Reference/>` element processing just before calculating the digest. And log `canonOutput` variable from `DigestMethod`?
https://github.com/PeculiarVentures/xmldsigjs/blob/master/src/signed_xml.ts#L360

It could help to compare canonicalized results from different applications and catch the problem

**Comment by linkurzweg** [comment]
@microshine Thank you very much for replying so quickly! I am not sure I understand. I ran `xmlsec1` with the `--store-references` option, but I am not sure which part of the printed output I'd need. Also, how exactly would I log the `canonOutput` variable from my code? Unfortunately I don't really have a clue about how XML signature verification works.

In the meantime I have asked If I could make an exception and be allowed to send you two example files for testing. If I get the permission, can I send them to the email account in your GitHub profile? Thanks again!

**Comment by microshine** [comment]
> can I send them to the email account in your GitHub profile?

Yes, you can. Please send me files to microshine@mail.ru

**Comment by linkurzweg** [comment]
@microshine Just got the permission and sent you the files. Thanks! :)

**Comment by microshine** [comment]
@linkurzweg Thank you for the files you've shared. I found the problem and going to understand the reason for it and fix it.

The problem is that canonicalized reference doesn't have the new line char at the end of the XML document (`xmlsec1` has that char).

**Comment by linkurzweg** [comment]
@microshine Thank you very much for looking into it! So there's nothing I need to do on my side, just wait for an update on the package?

**Comment by microshine** [comment]
yes

**Comment by microshine** [comment]
Here is an example of the problem

```js
var doc = new DOMParser().parseFromString(`<?xml version="1.0" encoding="UTF-8" standalone="yes"?><Root/>\n`, "application/xml");

new XMLSerializer().serializeToString(doc);
// '<?xml version="1.0" encoding="UTF-8" standalone="yes"?><Root/>'
```

As you can see `\n` doesn't present in the serialized document.

I'm looking for a solution how to solve it

**Comment by linkurzweg** [comment]
Very happy to read that. Kind of crazy that a new line char is enough to break the signature verification.

**Comment by linkurzweg** [comment]
@microshine Happy new year! Any news on this issue? Sorry for pestering you, just asking because I'm under a bit of pressure regarding my project. Thank you! :)

**Comment by microshine** [comment]
@linkurzweg Thank you for the congratulation. Happy new year too!

I've just published the new version of `xmldsigjs@2.2.0`. It fixes the issue.

```js
const sigFile = fs.readFileSync("/path/to/file.sign.xml", { encoding: "utf-8" });
const dataFile = fs.readFileSync("/path/to/file.xml"); // You may use any file format (not only XML)

const sig = new xmldsig.SignedXml();
sig.LoadXml(sigFile);
const ok = await sig.Verify({
    content: dataFile,
});
assert.ok(ok); // success
```

Please try  this version

**Comment by linkurzweg** [comment]
@microshine I just tried it and can confirm it works now! Thank you so much :)
